gcloud-java-core/src/main/java/com/google/gcloud/PageImpl.java

```diff
@@ -35,6 +38,33 @@
     Page<T> nextPage();
   }
 
+  static class PageIterator<T> extends AbstractIterator<T> {
+
+    private Iterator<T> currentPageIterator;
+    private Page<T> currentPage;
+
+    PageIterator(Page<T> currentPage) {
+      this.currentPageIterator = currentPage.values().iterator();
+      this.currentPage = currentPage;
+    }
+
+    @Override
+    protected T computeNext() {
+      if (currentPageIterator.hasNext()) {
+        return currentPageIterator.next();
+      }
+      Page<T> nextPage = currentPage.nextPage();
+      if (nextPage != null) {
+        currentPage = nextPage;
+        currentPageIterator = currentPage.values().iterator();
```

Comment:
Yes, I think it is possible (I know for sure that it is possible in the case of Datastore but this does not use `Page` right now).

Consider this case: listing with a filter.
Service use a cursor to jump to the start place and then iterate over the result to find values that match the filter criteria. At some point the service might give up before any row matched the criteria returning an empty result but with a cursor to allow continuing the listing from the point it stopped.

I admit that this should not be common but as possibility exists any reason for not supporting it?